### PR TITLE
Fix Architect.Construct to avoid throwing incorrect RangeErrors

### PR DESCRIPTION
**Pull request by AkashSamlal**
Addressing #78 issue

### CODE REVIEW

File: src/architecture/architect.js [hunk 1]
```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
...
+*
+* // Input: 2 neurons, 4 Hidden layers: 10 neurons, Output: 1 neuron
+* let myPerceptron = new architect.Perceptron(2, 10, 10, 10, 10, 1);
+*
+* @returns {Network} Feed forward neural network
+*/
+Perceptron: function () {
+  // Convert arguments to Array
+  // Coerce arguments to be Array-like w/ .call
+  var sizes = Array.prototype.slice.call(arguments); //type-casting/coercion arguments into array
```
Comment (dan-ryan) [inline src/architecture/architect.js]: You can use the newer Array.from instead these days.
https://stackoverflow.com/questions/36995617/convert-array-like-objects-array-prototype-slice-or-array-from

File: src/architecture/architect.js [hunk 2]
```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
...
-    }
+  // Create a list of nodes/groups
+  var groups = [];
+  groups.push(new Group(sizes[0], "input"));
 
-    for (i = 0; i < backconnections; i++) {
-      network.mutate(methods.mutation.ADD_BACK_CONN);
-    }
+  for (let i = 1; i < sizes.length; i++) {
+    //if(i === sizes.length-1) {
```
Comment (dan-ryan) [inline src/architecture/architect.js]: Commented out code.

File: src/architecture/architect.js [hunk 3]
```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
...
+}
 };
 
 module.exports = architect;
+
+
+
+
+
+
```
Comment (dan-ryan) [inline src/architecture/architect.js]: Pointless empty space.

File: src/architecture/architect.js [hunk 4]
```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
...
+  //layer hidden layer output
+  if (sizes.length < 3) {
+    throw new Error('You have to specify at least 3 layers');
+  }
 
-    for (i = 0; i < connections - hidden; i++) {
-      network.mutate(methods.mutation.ADD_CONN);
-    }
+  // Create a list of nodes/groups
+  var groups = [];
```
Comment (dan-ryan) [inline src/architecture/architect.js]: let

File: src/architecture/architect.js [hunk 5]
```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
...
-        inputs.push(nodes[i]);
-        nodes.splice(i, 1);
-      }
-    }
+  // Determine input and output nodes
+  var inputs = [];
+  var outputs = [];
+  for (i = nodes.length - 1; i >= 0; i--) {
+
+   if (nodes[i].type === 'output' || nodes[i].connections.in.length && nodes[i].connections.out.length + nodes[i].connections.gated.length === 0) {
```
Comment (dan-ryan) [inline src/architecture/architect.js]: A new null check is added here. But is it `nodes[i].connections.in.length` that is nullable or is it `nodes[i].connections.in`? Does the same check need to happen for `nodes[i].connections.out`?

Comment (christianechevarria) [reply src/architecture/architect.js]: Hey Dan! Good catch! There probably should be an equivalent test for the input case where `else if (nodes[i].type === 'input' || nodes[i].connections.out.length && !nodes[i].connections.in.length)`

File: src/architecture/architect.js (hunk 1 above)
Comment (christianechevarria) [reply src/architecture/architect.js]: Nice 🚀 ! Thanks for sending this over Dan, we're going to use and abuse this 😆

File: src/architecture/architect.js [hunk 6]
```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
...
+   let outputs = groups.map(function(group) {
+    let outs = group.nodes.map(function(node) {
+      if(node.type === "output") return node;
+    })
+    if(outs && outs.length > 1) return (outs)
+  })
+
+  console.log("# of inputs: " + inputs.length, inputs)
++    console.log("# of hidden: " + hidden.length, hidden)
++    console.log("# of outputs: " + outputs.length, outputs)
```
Comment (christianechevarria) [inline src/architecture/architect.js]: Stray `console.log`

File: src/architecture/architect.js [hunk 7]
```diff
@@ -5,562 +5,596 @@ var Layer = require('./layer');
...
+  // Create a list of nodes/groups
+  var groups = [];
+  groups.push(new Group(sizes[0], "input"));
 
-    for (i = 0; i < backconnections; i++) {
-      network.mutate(methods.mutation.ADD_BACK_CONN);
-    }
+  for (let i = 1; i < sizes.length; i++) {
+    //if(i === sizes.length-1) {
+      console.log("size " + (i+1) + " " + sizes[i])
```
Comment (christianechevarria) [inline src/architecture/architect.js]: Stray console.log

**Comment by luiscarbonell** [comment]
@AkashSamlal I'm going to close the PR for now, it's gone a bit stale - if you get back around to it, let me know - I would love to add it 🤙

**Pull request closed without merging**